### ISSUE TITLE
Poor calibration when using sample_weights

### ISSUE THREAD

**Issue by timchap92** [microsoft/LightGBM#5958]
## Description

I am using LightGBM to model an imbalanced binary classication problem (positive label rate is around 1:200). The data has been asymmetrically subsampled (90% of negative class observations are removed) to reduce data volume. I have accordingly set the sample weight for all negative class observations to 10.

The issue is that LightGBM exhibits very poor calibration on the holdout (test/validation) data when sample weights are used for this kind of imbalanced problem. In particular, for minority positive labels, LightGBM's predictions on the test data are consistently 10%-20% higher than for the train data (despite the fact that the data comes from the same distribution).

When the model is aggressively tuned to under-fit (e.g. by setting high values for `min_data_in_leaf` or `min_sum_hessian_in_leaf`), this problem disappears. For this reason, my interpretation is that this is a kind of overfitting. But the hyperparameters which minimise the objective (i.e. log loss) seem to consistently result in poor calibration.

Do you have any recommendations to mitigate this issue and ensure that predictions are well calibrated? I currently tune a large number of hyperparameters, including any that seem to be related to overfitting (`max_depth`, `learning_rate`, `path_smooth`, `min_data_in_leaf`, `min_sum_hessian_in_leaf`, `min_split_gain`, `num_leaves`, `num_iterations`, etc.)

## Reproducible example
```python
import random
import numpy as np
from lightgbm import LGBMClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.metrics import log_loss

random.seed(0)
np.random.seed(0)
X, y = make_classification(n_samples=10**6, n_informative=10, weights=[0.99, 0.01])

# Subsample negative class
subsample_rate = 0.1
keep_sample = np.logical_or.reduce([y == 1, np.random.rand(y.shape[0]) < subsample_rate])
X = X[keep_sample]
y = y[keep_sample]

w = np.ones_like(y) / subsample_rate
w[y == 1] = 1

X_train, X_test, y_train, y_test, w_train, w_test = train_test_split(X, y, w)

clf = LGBMClassifier(random_state=0)
clf.fit(X_train, y_train, sample_weight=w_train)

p_train = clf.predict_proba(X_train)[:, 1]
p_test = clf.predict_proba(X_test)[:, 1]

print(f'Train loss: {log_loss(y_train, p_train, sample_weight=w_train)}')
print(f'Test loss: {log_loss(y_test, p_test, sample_weight=w_test)}')

print(f'Mean pred train: {(p_train * w_train).sum() / w_train.sum()}')
print(f'Mean label train: {(y_train * w_train).sum() / w_train.sum()}')

print(f'Mean pred test: {(p_test * w_test).sum() / w_test.sum()}')
print(f'Mean label test: {(y_test * w_test).sum() / w_test.sum()}')
```

Output:

```
Train loss: 0.040031721581402745
Test loss: 0.0502729675920403
Mean pred train: 0.015120217715055714
Mean label train: 0.015096301218636456
Mean pred test: 0.017147299168392718
Mean label test: 0.014936516793044344
```

We observe that the average prediction on test data is significantly higher (~15%) than the average prediction on train data, even though the two distributions are the same. In this example there is no hyperparameter optimisation, but in other instances i have observed exactly the same over-prediction behaviour despite an extensive hyperparameter search (via Optuna). It seems that optimising the test/validation objective does not necessarily guarantee well-calibrated predictions when sample weights are used.

## Environment info

LightGBM version or commit hash: 3.3.5

Command(s) you used to install LightGBM

```shell
poetry add lightgbm
```

**Comment by timchap92** [comment]
Update: I wanted to justify this claim more thoroughly:
>It seems that optimising the test/validation objective does not necessarily guarantee well-calibrated predictions when sample weights are used.

To this end, I extended the example above with a quick-and-dirty hyperparameter search:
```python
...

grid = ParameterGrid(
    {
        'min_child_weight': np.logspace(-1, 2, num=4),
        'learning_rate': np.logspace(-2, 0, num=4),
        'num_iterations': np.logspace(2, 3, num=4)
    }
)
test_loss = []
test_calib = []
for params in grid:
    clf = LGBMClassifier(random_state=0, **params)

    ...

    test_loss.append(log_loss(y_test, p_test, sample_weight=w_test))
    test_calib.append(
        (p_test * w_test).sum() / w_test.sum()
        / ((y_test * w_test).sum() / w_test.sum())
    )

plt.scatter(test_loss, test_calib)
plt.xlim(0.04, 0.08)
plt.ylim(0.95, 1.5)
plt.xlabel('Log loss')
plt.ylabel('Calibration')
```

This gives:
![image](https://github.com/microsoft/LightGBM/assets/13381054/21269053-e84d-4d0c-9bdb-1c82beb0be37)

This reinforces (I think) that there is a clear tendency for poorly calibrated estimations when optimizing for test log loss.

**Comment by tim-voodoo** [comment]
Let me know if further explanation is needed :)

**Comment by timchap92** [comment]
Let me know if you need any further explanation :)
